tls/s2n_x509_validator.c

```diff
@@ -357,40 +347,112 @@ S2N_RESULT s2n_x509_validator_validate_cert_chain(struct s2n_x509_validator *val
+    }
+
+    RESULT_GUARD_OSSL(X509_STORE_CTX_init(validator->store_ctx, validator->trust_store->trust_store, leaf,
+            validator->cert_chain_from_wire), S2N_ERR_INTERNAL_LIBCRYPTO_ERROR);
+
+    validator->state = CERT_CHAIN_PROCESSED;
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_x509_validator_verify_cert_chain(struct s2n_x509_validator *validator, struct s2n_connection *conn) {
+    RESULT_ENSURE(validator->state == CERT_CHAIN_PROCESSED, S2N_ERR_INVALID_CERT_STATE);
+
+    if (validator->skip_cert_validation) {
+        return S2N_RESULT_OK;
+    }
+
+    X509_VERIFY_PARAM *param = X509_STORE_CTX_get0_param(validator->store_ctx);
+    X509_VERIFY_PARAM_set_depth(param, validator->max_chain_depth);
+
+    uint64_t current_sys_time = 0;
+    conn->config->wall_clock(conn->config->sys_clock_ctx, &current_sys_time);
+
+    /* this wants seconds not nanoseconds */
+    time_t current_time = (time_t)(current_sys_time / 1000000000);
+    X509_STORE_CTX_set_time(validator->store_ctx, 0, current_time);
+
+    int verify_ret = X509_verify_cert(validator->store_ctx);
+    if (verify_ret <= 0) {
+        int ossl_error = X509_STORE_CTX_get_error(validator->store_ctx);
+        switch (ossl_error) {
+            case X509_V_ERR_CERT_HAS_EXPIRED:
+                RESULT_BAIL(S2N_ERR_CERT_EXPIRED);
+            default:
+                RESULT_BAIL(S2N_ERR_CERT_UNTRUSTED);
         }
+    }
+
+    validator->state = VALIDATED;
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_x509_validator_read_leaf_info(struct s2n_connection *conn, uint8_t *cert_chain_in, uint32_t cert_chain_len,
+        struct s2n_pkey *public_key, s2n_pkey_type *pkey_type, s2n_parsed_extensions_list *first_certificate_extensions) {
+    struct s2n_blob cert_chain_blob = {.data = cert_chain_in, .size = cert_chain_len};
+    DEFER_CLEANUP(struct s2n_stuffer cert_chain_in_stuffer = { 0 }, s2n_stuffer_free);
+
+    RESULT_GUARD_POSIX(s2n_stuffer_init(&cert_chain_in_stuffer, &cert_chain_blob));
+    RESULT_GUARD_POSIX(s2n_stuffer_write(&cert_chain_in_stuffer, &cert_chain_blob));
+
+    struct s2n_blob asn1_cert = { 0 };
+    RESULT_GUARD(s2n_x509_validator_read_asn1_cert(&cert_chain_in_stuffer, &asn1_cert));
 
-        validator->state = VALIDATED;
+    RESULT_ENSURE(s2n_asn1der_to_public_key_and_type(public_key, pkey_type, &asn1_cert) == 0,
+            S2N_ERR_CERT_UNTRUSTED);
+
+    /* certificate extensions is a field in TLS 1.3 - https://tools.ietf.org/html/rfc8446#section-4.4.2 */
+    if (conn->actual_protocol_version >= S2N_TLS13) {
+        s2n_parsed_extensions_list parsed_extensions_list = { 0 };
+        RESULT_GUARD_POSIX(s2n_extension_list_parse(&cert_chain_in_stuffer, &parsed_extensions_list));
+
+        /* RFC 8446: if an extension applies to the entire chain, it SHOULD be included in the first CertificateEntry */
+        *first_certificate_extensions = parsed_extensions_list;
+    }
```

Comment:
Not directly related to your change, and I know we already only processed the extensions from the first certificate, but it still feels like very wrong behavior to me. Can we put a comment somewhere (either here or down where the extensions are processed) making this behavior very clear and mentioning that it's an s2n-tls choice not specified by the RFC?

Reply:
I added a comment where the extensions are processed: 2b5f26ea25abbfbdd2967bf021f057ac88366a0a

I also moved the existing RFC SHOULD comment to here.